cfps/CFP-041-clang-format-for-bpf-formatting.md

```diff
@@ -0,0 +1,81 @@
+* Add these changes to a `.git-blame-ignore` file so the commit doesn't mess up the blame history.
+* Change CI to use clang-format to detect formatting issues and switch off formatting checks in checkpatch.pl.
+* Backport these changes to older, still maintained branches to minimize merge conflicts during backporting.
+* Bypass the normal CODEOWNERS review process, instead having the committee review and approve the PR(s) needed for this change.
+
+### The Committee
+
+Even when trying to get clang-format as close to the existing format, there are still a number of decisions that need to be made. Our current formatting is internally inconsistent having multiple standards for indenting, spacing, alignment, etc. between and within files. These aspects where previously not enforced by checkpatch.pl and up to the individual contributor to decide. However, clang-format requires us to make these decisions.
+
+code formatting / code style is infamous for being a topic that can lead to heated debates and bikeshedding. Unfortunately chances are slim that we can make every contributor happy. We propose to create a small committee we trust to make these decisions and to given them the power to make these decisions. The committee will consist of Dylan Reimerink being the committer to work on this primarily and two other people yet to be chosen.
+
+### Existing format and `.clang-format` config
+
+The starting point of this work will be the existing format. We already have a `.clang-format` config file in the repository which has been created in the past to approximate the existing formatting style. The committee will review the changes made by clang-format and decide on the best settings to use.
+
+### Initial re-formatting commit
+
+file so the commit doesn't mess up the blame history.codebase and commit the changes. This will be a single commit to avoid polluting the git history with formatting changes. We will add this commit to a `.git-blame-ignore` which is recognized by both the git cli tool and GitHub to ignore this commit when showing blame information to minimize the impact.
+
+### Tooling changes
+
+We propose to update the cilium-builder docker image which builds our clang to also provide `clang-format`. We will add make targets to run `clang-format` in-place (directly update files) or output any suggested changes to stdout.
+
+In CI we will run one of these targets to detect if a committer properly formatted their code (clang-format should give no suggestions).
+
+We will modify the checkpatch script in the checkpatch docker iamge, disabling the formatting checks now covered by clang-format and leaving all other checks enabled.
+
+### Backporting
+
+We propose to backport the config, tooling and CI changes to older branches we still maintain. Then run clang-format on these branches to update their formatting in the same way and adding the commit in a `.git-blame-ignore` file.
```

Comment:
Do we use the same cilium-builder or a different one? Does this impact the version of clang-format we use?

Reply:
Good question actually, we can't use newer config files on older clang-format instances since it will likely have settings the older tool doesn't know. But older configs do work with newer versions.

We could easily use the newest available clang-format on all branches. While I don't expect major differences in formatting from version to version, it is certainly possible. So using the latest version everywhere would reduce backport pain the most.